src/engine/strat_engine/filesystem.rs

```diff
@@ -47,11 +56,35 @@ impl StratFilesystem {
         }
     }
 
-    pub fn check(&self, dm: &DM) -> EngineResult<FilesystemStatus> {
+    /// check if filesystem is getting full and needs to be extended
+    /// TODO: deal with the thindev in a Fail state.
+    pub fn check(&mut self, dm: &DM) -> EngineResult<FilesystemStatus> {
         match try!(self.thin_dev.status(dm)) {
-            ThinStatus::Good((_mapped, _highest)) => {
-                // TODO: check if filesystem is getting full and might need to
-                // be extended (hint: use statfs(2))
+            ThinStatus::Good((_, _)) => {
+                let mount_point = try!(self.get_mount_point());
+                let (fs_total_bytes, fs_total_used_bytes, fs_block_size) =
+                    try!(fs_usage(&mount_point));
+                let free_bytes = Bytes((*fs_total_bytes - *fs_total_used_bytes) / *fs_block_size);
+                // block sizes for the filesystem and thin dev are different, compare the space
+                // as bytes
+                if *free_bytes < *FILESYSTEM_LOWATER * *DATA_BLOCK_SIZE * SECTOR_SIZE as u64 {
+                    // current filesystem blocks to extend
+                    let extend_fs_blocks = *fs_total_bytes / *fs_block_size;
+                    // Extend the thin_dev so it will have enough space for the number
+                    // of sectors in the filesystem blocks for XFS grow request.
+                    match self.thin_dev
+                              .extend(dm,
+                                      Sectors(extend_fs_blocks * *fs_block_size /
+                                              SECTOR_SIZE as u64)) {
+                        Ok(_) => {}
+                        Err(_) => return Ok(FilesystemStatus::ThinDevExtendFailed),
+                    }
+                    match xfs_growfs(&mount_point,
+                                     self.extend_size(DataBlocks(extend_fs_blocks))) {
```

Comment:
I think that we can not be sure that filesystem blocks and our DataBlocks are the same, so this is a bug. If we can be sure, then there are a lot of things we can simplify, but I think not.